### PR TITLE
Expand documentation on extensions

### PR DESCRIPTION
**Pull request by jlaehne**
### Description of the change
- Expand the documentation related to HyperSpy extensions by adding a dedicated section to the signal/signal_basics page
- Adapt notes about the coming version 2
- A checkpoint has been added to #2996, so that we do not forget to add InterSphinx mapping to HoloSpy and EleXSpy once their documentation is available

### Progress of the PR
- [x] Change implemented (can be split into several points),
- [x] ready for review.

### CODE REVIEW

File: doc/intro.rst [hunk 1]
```diff
@@ -17,17 +17,21 @@ multidimensionality.
   components and widgets.
 
 
-.. _hyperspy_extensions-label:
+The functionality of HyperSpy can be extended by external packages, e.g. to
+implement features for analyzing a particular sort of data (usually related to a
+specific set of methods). A `list of packages that extend HyperSpy
```
Comment (hakonanes) [inline doc/intro.rst]:
```suggestion
specific set of experimental methods). A `list of packages that extend HyperSpy
```

File: doc/intro.rst [hunk 2]
```diff
@@ -17,17 +17,21 @@ multidimensionality.
...
-:ref:`writing_extensions-label`. For a list of packages that extend HyperSpy
-follow `this link <https://github.com/hyperspy/hyperspy-extensions-list>`_.
 
-.. note::
-    From version 2.0, HyperSpy will be split into a core package (HyperSpy)
-    that will provide the common infrastructure and a number of HyperSpy
-    extensions specialized in the analysis of different types of data.
+.. versionchanged:: 2.0
+    HyperSpy was split into a core package (HyperSpy) that provides the common
+    infrastructure for multidimensional datasets and a number of `HyperSpy
```
Comment (hakonanes) [inline doc/intro.rst]: This makes it sound like pyxem, kikuchipy, and LumiSpy was split from HyperSpy. At least for kikuchipy, this is not the case. Can it be rephrased? For example:

"HyperSpy was split into a core package (HyperSpy) that provides the common infrastructure for multidimensional datasets and a dedicated IO package RosettaSciIO. HyperSpy is extended by packages that focuses on analysis of certain types of data."

File: doc/user_guide/signal/signal_basics.rst [hunk 3]
```diff
@@ -133,111 +129,149 @@ example we create a Signal2D instance from a 2D numpy array:
...
+    |      hologram      |                     |   HologramImage    | holospy  |
     +--------------------+---------------------+--------------------+----------+
 
-.. warning::
-    From version 2.0 HyperSpy will no longer ship
-    :py:class:`~.api.signals.BaseSignal` subclasses that are specific to a
-    particular type of data (i.e. with non-empty ``signal_type``). All those
-    signals currently distributed with HyperSpy will be moved to new
-    packages.
+Object oriented functions provided by the extensions will be available after
```
Comment (hakonanes) [inline doc/user_guide/signal/signal_basics.rst]: This information is not clear to me. Which functions from extensions are available after `import hyperspy.api as hs`? Do you mean that an extension signal loaded with `s = hs.load("signal.h5", signal_type="EBSD")` will have additional class methods not provided by HyperSpy?

**Review by hakonanes — COMMENTED** [review]
These are great changes that give a better understanding of extensions.

I've added a few questions.

File: doc/intro.rst (hunk 2 above)
Comment (jlaehne) [reply doc/intro.rst]: Well the changing part is that some signal classes have been moved to extensions, but I'll phrase that more explicitly.

File: doc/user_guide/signal/signal_basics.rst (hunk 3 above)
Comment (jlaehne) [reply doc/user_guide/signal/signal_basics.rst]: Yes, for example for an EBSD signal, you provide the `s.extract_grid()` method. I'll try to rephrase it to make it more clear.

File: doc/intro.rst (hunk 2 above)
Comment (hakonanes) [reply doc/intro.rst]: Great, thanks for making the change.

File: doc/user_guide/signal/signal_basics.rst (hunk 3 above)
Comment (hakonanes) [reply doc/user_guide/signal/signal_basics.rst]: Thanks for updating. Perhaps the final sentence in that paragraph could be a recommendation to visit the extensions' references for lists of these functions and methods?

File: doc/user_guide/signal/signal_basics.rst [hunk 4]
```diff
@@ -133,111 +129,150 @@ example we create a Signal2D instance from a 2D numpy array:
...
-    packages.
+Functions provided by the extensions that directly operate on the data types
+defined in a registered subclass (object oriented functions) will be available
+after loading the ``hyperspy.api``. However, the packages might also provide some
+additional utilities, which require them to be separately loaded into the python
+namespace.
+
+When :ref:`loading data <loading_files>`, a specific ``signal_type`` will be
+set automatically if it is recognized by the file reader and the corresponding
+extension is installed on your system.
```
Comment (ericpre) [inline doc/user_guide/signal/signal_basics.rst]: This is not very clear, maybe this is better?

```suggestion
When :ref:`loading data <loading_files>`, the ``signal_type`` will be
set automatically by the file reader, as defined in ``rosettasciio``. If the
extension providing the specialised signal subclass corresponding to the given ``signal_type`` is installed,  :py:func:`~.api.load` will return the subclass from the hyperspy extension, otherwise a warning will be raised to explain that no registered signal class can be assigned to the given ``signal_type``.
```

Comment (jlaehne) [reply doc/user_guide/signal/signal_basics.rst:232]: I don't know if all file readers will raise a warning, or if some might just resort to Signal1D.

File: doc/user_guide/signal/signal_basics.rst [hunk 5]
```diff
@@ -133,111 +129,155 @@ example we create a Signal2D instance from a 2D numpy array:
...
 
-.. warning::
-    From version 2.0 HyperSpy will no longer ship
-    :py:class:`~.api.signals.BaseSignal` subclasses that are specific to a
-    particular type of data (i.e. with non-empty ``signal_type``). All those
-    signals currently distributed with HyperSpy will be moved to new
-    packages.
+Functions provided by the extensions that directly operate on the data types
+defined in a registered subclass (object oriented functions) will be available
+after loading the ``hyperspy.api``. However, the packages might also provide some
```
Comment (ericpre) [inline doc/user_guide/signal/signal_basics.rst]: I don't understand what does it mean. Do you mean that `hs.load` will return signal class provided by extension and therefore the methods implemented in extension will be available without explicitly loading their namespace?
I am not sure that "object oriented functions" is a good wording.

Comment (jlaehne) [reply doc/user_guide/signal/signal_basics.rst]: Yes, that is what I mean and the term "object oriented programming" is what I learned for this paradigm where a data class ships its own methods that can manipulate the data back when I learned programming at uni. Do you have a better proposal?

Comment (ericpre) [reply doc/user_guide/signal/signal_basics.rst]: Yes, "object oriented programming" is good - it was the word "function" in "object oriented functions" which was unclear to me.
Should this paragraph comes after the one below where loading data is introduced?

I think that the message here is that `hyperspy.api.load` will returns signal object implemented in extension. When saying that "registered subclass (object oriented functions) will be available after loading the ``hyperspy.api``" it sounds like the class will be added the hyperspy namespace.

File: doc/user_guide/signal/signal_basics.rst [hunk 6]
```diff
@@ -133,111 +129,155 @@ example we create a Signal2D instance from a 2D numpy array:
...
+extension providing the corresponding signal subclass is installed,
+:py:func:`~.api.load` will return the subclass from the hyperspy extension,
+otherwise a warning will be raised to explain that
+no registered signal class can be assigned to the given ``signal_type``.
+
+Functions/methods provided by the extensions that directly operate on the data
+types (thus being object oriented) defined in a registered subclass will be available
+after loading the ``hyperspy.api``. However, the packages might also provide some
+additional utilities, which require them to be separately loaded into the python
+namespace. Check the user guides of the `HyperSpy extensions
```
Comment (ericpre) [inline doc/user_guide/signal/signal_basics.rst]: I think that we should focus on what returns `hs.load` rather than importing `hyperspy`. Returning objects provided by other libraries is very common (for example, function that return numpy array), what is special here is that `hs.load` return subclasses of `BaseSignal` implemented in different libraries.
What's about the following?
```suggestion
Since the :py:func:`~.api.load` can return domain specific signal objects (e.g. ``EDSSEMSpectrum`` from ``EleXSpy``) provided by extensions, the corresponding functionalities (so-called `method` of `object` in object oriented programming, e.g. ``EDSSEMSpectrum.get_lines_intensity``) implemented in signal classes of the extension can be accessed readily.
To use additional functionalities implemented in extension but not as method of the signal class, the extensions will need to be imported explicitly (e.g. ``import EleXSpy``). Check the user guides of the `HyperSpy extensions
```

**Review by ericpre — APPROVED** [review]
